[scheduler] Support preemption of pods using ReadWriteOncePod PVCs

chrishenzie opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds support for preemption of pods using `ReadWriteOncePod` PVCs. This is a required feature for [beta graduation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode#beta).

#### Which issue(s) this PR fixes:
Fixes #103132

#### Does this PR introduce a user-facing change?
```release-note
Adds scheduler preemption support for pods using `ReadWriteOncePod` PVCs
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode
```

#### Testing
This feature can be tested the following ways:
```sh
# Unit tests.
go test ./pkg/scheduler/framework/... ./pkg/scheduler/internal/...

# Integration tests.
go test -v -run TestUnschedulablePodBecomesSchedulable/scheduled_pod_uses_read-write-once-pod_pvc ./test/integration/scheduler/filters
go test -v ./test/integration/scheduler/preemption/ -run TestReadWriteOncePodPreemption

# And E2E tests.
make WHAT=test/e2e/e2e.test
_output/local/bin/linux/amd64/e2e.test --kubeconfig=${HOME}/.kube/config -ginkgo.focus='Feature:ReadWriteOncePod'
```

/sig storage
/sig scheduling
/cc @msau42
/cc @jsafrane
/cc @alculquicondor

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +43,90 @@ type VolumeRestrictions struct {
+	if s == nil {
+		return
+	}
+	if podInfo == nil {
+		return
+	}
+	if multiplier == 0 {
+		return
+	}
```
are these possible?

chrishenzie replied:
No, removing

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +43,90 @@ type VolumeRestrictions struct {
+}
+
+func (s *preFilterState) hasConflict(podInfo *framework.PodInfo) bool {
+	if s == nil {
+		return false
+	}
+	if podInfo == nil {
+		return false
+	}
```
if not reachable, just remove these lines. Otherwise we should have test coverage for those cases.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +43,90 @@ type VolumeRestrictions struct {
+			continue
+		}
+
+		pvcName := volume.PersistentVolumeClaim.ClaimName
+		if s.readWriteOncePodPVCs.Has(pvcName) {
```
```suggestion
		if s.readWriteOncePodPVCs.Has(volume.PersistentVolumeClaim.ClaimName) {
```

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +43,90 @@ type VolumeRestrictions struct {
+	}
+
+	copy := &preFilterState{}
+	copy.readWriteOncePodPVCs = s.readWriteOncePodPVCs.Clone()
```
only clone if `AddPod` and `RemovePod` change these values. I don't think they do, so just a shallow copy is enough for this field.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +43,90 @@ type VolumeRestrictions struct {
+}
+
+func clonePodMap(m map[string]*v1.Pod) map[string]*v1.Pod {
+	n := make(map[string]*v1.Pod)
```
```suggestion
	n := make(map[string]*v1.Pod, len(m))
```

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("error reading %q from cycleState: %w", preFilterStateKey, err)
```
```suggestion
		return nil, fmt.Errorf("reading %q from cycleState: %w", preFilterStateKey, err)
```

It's an already an error, so no need to say that in the message.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
-// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as UnschedulableAndUnresolvable.
-// TODO(#103132): Mark pod as Unschedulable and add preemption logic.
-func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(ctx context.Context, pod *v1.Pod) *framework.Status {
+func (pl *VolumeRestrictions) readWriteOncePodPVCsForPod(ctx context.Context, pod *v1.Pod) ([]*v1.PersistentVolumeClaim, error) {
```
why not make this return a set already?

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
-	if pl.enableReadWriteOncePod {
-		return nil, pl.isReadWriteOncePodAccessModeConflict(ctx, pod)
+	s, err := pl.calPreFilterState(ctx, pod)
+	if err != nil {
```
one of the failures could be for PVC NotFound, in which case we should return `UnschedulableAndUnresolvable`, instead of error.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+	for _, p := range pods {
+		for _, volume := range p.Spec.Volumes {
+			// The current pod cannot be in conflict with itself.
+			if p.Name == pod.Name {
+				continue
+			}
```
This wouldn't happen, as the lister contains the scheduled pods.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -168,19 +366,18 @@ func (pl *VolumeRestrictions) PreFilterExtensions() framework.PreFilterExtension
-		if v.GCEPersistentDisk == nil && v.AWSElasticBlockStore == nil && v.RBD == nil && v.ISCSI == nil {
-			continue
-		}
+func (pl *VolumeRestrictions) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
```
Add a comment on top about the new logic

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+
+// Checks if scheduling the pod onto this node would cause any conflicts with
+// existing volumes.
+func satisfyExistingVolumes(pod *v1.Pod, nodeInfo *framework.NodeInfo) bool {
```
Can't think of a better name, but do you have other alternatives?

chrishenzie replied:
I changed it to `satisfyVolumeConflicts` but I don't really feel great about that either

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+	if state == nil {
+		return framework.NewStatus(framework.Success)
+	}
+	if pod == nil {
+		return framework.NewStatus(framework.Error)
+	}
```
not reachable

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+	reasons := []string{}
+
+	for _, conflictingPod := range state.conflictingPods {
+		conflictingPriority := corev1helpers.PodPriority(conflictingPod)
+		if conflictingPriority >= priority {
```
why are you checking priority?

The preemption plugin already handles priority when simulating removal of pods.

chrishenzie replied:
I didn't know, thought I might need to do this. From my understanding it sounds like (at a high-level):
- if PVC not found, return `UnschedulableAndUnresolvable`
- if there are conflicting pods, return `Unschedulable` (regardless of priority)
- else return success

alculquicondor replied:
correct

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -118,45 +194,167 @@ func haveOverlap(a1, a2 []string) bool {
+
+	priority := corev1helpers.PodPriority(pod)
+	code := framework.Success
+	reasons := []string{}
```
what't the point of having multiple reasons with the same value? :)

alculquicondor commented on pkg/scheduler/framework/types.go:
```diff
@@ -409,10 +409,6 @@ type NodeInfo struct {
 
-	// PVCRefCounts contains a mapping of PVC names to the number of pods on the node using it.
-	// Keys are in the format "namespace/name".
-	PVCRefCounts map[string]int
```
why are you removing these instead of reusing this information in Prefilter?

chrishenzie replied:
I thought I could rebuild the cache from scratch (but I see that's inefficient https://github.com/kubernetes/kubernetes/pull/114051#issuecomment-1322626693).

My cache kept references to the conflicting `*v1.Pod`s because I thought I'd need the conflicting pod's priority to determine which status to return. Since that isn't necessary, I think I can use this cache.

alculquicondor replied:
I think you can still make it work with just integers.

This can store how many pods reference the PVC.

Then, in the state data, when you remove a pod, you can substract the number of conflicts that the pod has.

alculquicondor reviewed (commented):
I didn't review tests yet.

chrishenzie committed:
feat: support preemption for pods using ReadWriteOncePod PVCs

PVCs using the ReadWriteOncePod access mode can only be referenced by a
single pod. When a pod is scheduled that uses a ReadWriteOncePod PVC,
return "Unschedulable" if the PVC is already in-use in the cluster.

To support preemption, the "VolumeRestrictions" scheduler plugin
computes cycle state during the PreFilter phase. This cycle state
contains the number of references to the ReadWriteOncePod PVCs used by
the pod-to-be-scheduled.

During scheduler simulation (AddPod and RemovePod), we add and remove
reference counts from the cycle state if they use any of these
ReadWriteOncePod PVCs.

In the Filter phase, the scheduler checks if there are any PVC reference
conflicts, and returns "Unschedulable" if there is a conflict.

This is a required feature for the ReadWriteOncePod beta. See for more context:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode#beta

chrishenzie commented:
I restructured the PR to go with an approach like https://github.com/kubernetes/kubernetes/pull/114051#issuecomment-1322626693. Please see the commit descriptions for each one for more context.

At a high-level, I've changed the caches to store references to the specific pods using each PVC. In `VolumeRestrictions`, we query this cache with the pod-to-be-scheduled's `ReadWriteOncePod` PVCs to build a cache of conflicting pods.

chrishenzie commented:
/milestone 1.27

chrishenzie commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -117,46 +164,127 @@ func haveOverlap(a1, a2 []string) bool {
+	if state == nil {
+		return nil
+	}
+	if len(state.conflictingPods) > 0 {
```
See here for a similar discussion: https://github.com/kubernetes/kubernetes/pull/114051#discussion_r1029589393, this cache does not include the pod-to-be-scheduled. 

Also confirmed this manually with debug logging (I added comments for context). These are collected after running the e2e test in this PR:
```
# First pod gets scheduled, has no conflicts
podToBeScheduled is: pod-65fbe1d7-0525-4909-afa2-0c4f953f00cc, conflictingPods are: []

# Second pod (with higher priority) gets scheduled and conflicts with first pod
podToBeScheduled is: pod-952eb976-f8f1-4377-8dfe-a2abf235664e, conflictingPods are: [pod-65fbe1d7-0525-4909-afa2-0c4f953f00cc]

# First pod is preempted, scheduling of second pod shows no more conflicts
podToBeScheduled is: pod-952eb976-f8f1-4377-8dfe-a2abf235664e, conflictingPods are: []

# First pod is re-created and conflicts with second pod
podToBeScheduled is: pod-65fbe1d7-0525-4909-afa2-0c4f953f00cc, conflictingPods are: [pod-952eb976-f8f1-4377-8dfe-a2abf235664e]

# Second pod is deleted, scheduling of first pod shows no more conflicts
podToBeScheduled is: pod-65fbe1d7-0525-4909-afa2-0c4f953f00cc, conflictingPods are: []
```

chrishenzie commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -117,46 +164,127 @@ func haveOverlap(a1, a2 []string) bool {
 
+// PreFilter computes and stores cycleState containing details for enforcing ReadWriteOncePod.
 func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) (*framework.PreFilterResult, *framework.Status) {
-	if pl.enableReadWriteOncePod {
```
Good catch, added this to `PreFilter`, `AddPod` and `RemovePod`

chrishenzie commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -168,19 +296,20 @@ func (pl *VolumeRestrictions) PreFilterExtensions() framework.PreFilterExtension
+		return framework.AsStatus(err)
+	}
+
+	if pl.enableReadWriteOncePod {
```
I think if we did this then we would no longer execute the existing `satisfyVolumeConflicts()` logic if `!pl.enableReadWriteOncePod`.

chrishenzie commented:
This doesn't appear to be working (preempting lower priority pods) in a multi-node environment, but does for single-node. Looking into this.

alculquicondor commented:
yeah..... We don't have support for preemption across the cluster, it's too costly.

So we only simulate preemption within a node and check if that makes the Pod schedulable in that node. If it doesn't, we put the pods back, and move to evaluate the next node.

chrishenzie commented on test/e2e/storage/testsuites/readwriteoncepod.go:
```diff
@@ -121,11 +131,18 @@ func (t *readWriteOncePodTestSuite) DefineTests(driver storageframework.TestDriv
 	})
 
-	ginkgo.It("should block a second pod from using an in-use ReadWriteOncePod volume", func() {
+	ginkgo.It("should preempt lower priority pods using ReadWriteOncePod volumes", func() {
```
I just demoed this scenario and it appears to be working.

First applied this manifest to create Pod1 (using ReadWriteOncePod PVC), Pod2, Pod3:
https://gist.github.com/chrishenzie/710881374ea852055d37e988cee9c523
```
$ kubectl get po
NAME                              READY   STATUS    RESTARTS   AGE
csi-mockplugin-76d64df48c-ddxnl   6/6     Running   0          17m
pod1                              1/1     Running   0          34s
pod2                              1/1     Running   0          34s
pod3                              1/1     Running   0          34s
```

Then applied this manifest which creates Pod4 which shares the PVC:
https://gist.github.com/chrishenzie/0fbbae2ba7bf1099a31fb718eb7cb80d
```
$ kubectl get po
NAME                              READY   STATUS    RESTARTS   AGE
csi-mockplugin-76d64df48c-ddxnl   6/6     Running   0          21m
pod2                              1/1     Running   0          4m33s
pod3                              1/1     Running   0          4m33s
pod4                              1/1     Running   0          41s
```

Pod1 was the only preempted pod because it was the only one with a conflicting PVC.

chrishenzie commented:
/retest

chrishenzie commented on test/e2e/storage/testsuites/readwriteoncepod.go:
```diff
@@ -141,20 +158,33 @@ func (t *readWriteOncePodTestSuite) DefineTests(driver storageframework.TestDriv
-		err = e2epod.DeletePodWithWait(l.cs, pod1)
-		framework.ExpectNoError(err, "failed to delete pod1")
+		// Wait for the first pod to be preempted and the second pod to start.
+		err = e2epod.WaitForPodToDisappear(l.cs, pod1.Namespace, pod1.Name, labels.Everything(), framework.Poll, f.Timeouts.PodStart)
```
Per this comment https://github.com/kubernetes/kubernetes/pull/114051/#issuecomment-1325639784, does this mean I'll need to add `nodeSelectors` to pod1 and pod2 to force preemption when this test runs in a multi-node environment?

alculquicondor replied:
Not necessarily. If the pod2 fits in the space that pod1 would leave, then preemption should work.

alculquicondor commented:
@kerthcet anything to add? I'll wait for your LGTM before approval

chrishenzie commented:
/hold
Waiting for 1.27

kerthcet reviewed (commented):
Really great, just some nits, I'll label with LGTM after squash.

chrishenzie commented:
Squashed and included bump in feature gate from alpha to beta.

kerthcet commented:
/retest

kerthcet commented:
/lgtm
Thanks @chrishenzie , for approval @alculquicondor

alculquicondor commented:
/priority important-longterm

alculquicondor commented:
/triage accepted

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +41,63 @@ type VolumeRestrictions struct {
+	// Names of the pod's volumes using the ReadWriteOncePod access mode.
+	readWriteOncePodPVCs sets.Set[string]
+	// Pods that are also using these ReadWriteOncePod volumes.
+	conflictingPods sets.Set[string]
```
I think this should be enough

```suggestion
	numConflictingPods int
```

alculquicondor commented:
Leave the graduation and E2E test to a separate PR.

chrishenzie commented on pkg/features/kube_features.go:
```diff
@@ -729,7 +729,9 @@ const (
 	// owner: @chrishenzie
+	// kep: https://kep.k8s.io/2485
 	// alpha: v1.22
+	// beta: v1.27
```
Will do

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +41,62 @@ type VolumeRestrictions struct {
+	}
+}
+
+func (s *preFilterState) hasConflict(podInfo *framework.PodInfo) bool {
```
shouldn't this return how many volumes are conflicting?

add a unit or integration test that exercises preemption when there are more than one conflict

alculquicondor replied:
I'm not convinced that this works, because in `calPreFilterState` you sum up the number of pods that conflict for each PV.

And in updateWithPod, you only count the pods if there is a conflict, independently from the number of pvcs that this pod is having a conflict with.

chrishenzie replied:
`calPreFilterState` had a bug where pods were double-counted when they share more than one conflicting PVC. I think if we now have an accurate count of `numConflictingPods` in PreFilter, then we should be able to increment and decrement it by 1 (pod) and not worry about the number of conflicting PVCs

chrishenzie replied:
See the most recent commit for context. Included is an integration test that caught this double-counting issue.

chrishenzie replied:
```
=== RUN   TestReadWriteOncePodPreemption/preempt_single_pod
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0

=== RUN   TestReadWriteOncePodPreemption/preempt_two_pods
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 2
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-2
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod-1
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 2

^  not sure why this happens yet but I think it's because the preemptor-pod is re-scheduled so it re-calculates the preFilterState

=== RUN   TestReadWriteOncePodPreemption/preempt_single_pod_with_two_volumes
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
scheduling pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/preemptor-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
adding pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 1
removing pod: preemption7ccd1c37-1360-4fb1-91e2-400c609ecbd5/victim-pod
numConflictingPods is 0
```

alculquicondor commented on pkg/scheduler/internal/cache/snapshot.go:
```diff
@@ -193,6 +196,7 @@ func (s *Snapshot) Get(nodeName string) (*framework.NodeInfo, error) {
-func (s *Snapshot) IsPVCUsedByPods(key string) bool {
-	return s.usedPVCSet.Has(key)
+// PodsUsingPVC returns the set of scheduled pod names using the PVC.
+func (s *Snapshot) PodsUsingPVC(key string) sets.Set[string] {
```
Is this enough?

```suggestion
func (s *Snapshot) PodsUsingPVC(key string) int {
```

chrishenzie replied:
We need the pod names so we don't double-count pods if they have more than one conflicting volume with the pod-to-be-scheduled. Once we have the unique set of conflicting pod names (in `calPreFilterState`) then we can just grab the `.Len()` for `numConflictingPods`.

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +41,64 @@ type VolumeRestrictions struct {
+func (s *preFilterState) updateWithPod(podInfo *framework.PodInfo, multiplier int64) {
+	conflicts := s.numConflictingPVCRefsForPod(podInfo)
+	if conflicts > 0 {
+		if multiplier > 0 {
+			s.numConflictingPVCRefs += conflicts
+		} else {
+			s.numConflictingPVCRefs -= conflicts
```
nit:
```suggestion
		s.numConflictingPVCRefs += multiplier * conflicts
```

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +41,64 @@ type VolumeRestrictions struct {
+}
+
+func (s *preFilterState) numConflictingPVCRefsForPod(podInfo *framework.PodInfo) int {
+	var conflicts int
```
nit (slightly more idiomatic):
```suggestion
	conflicts := 0
```

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -117,46 +165,135 @@ func haveOverlap(a1, a2 []string) bool {
+		return nil, framework.AsStatus(err)
+	}
+	cycleState.Write(preFilterStateKey, s)
+	return nil, nil
```
You could return `Unschedulable` in `PreFilter` in some scenarios to completely skip Filter, right?

Probably:

```golang
return nil, satisfyReadWriteOncePod(ctx, s)
```

chrishenzie replied:
Having trouble making this work:
```
=== RUN   TestReadWriteOncePodPreemption/preempt_single_pod
E1212 14:10:24.194140 1682349 framework.go:698] "Failed running RemovePod on PreFilter plugin" err="reading \"PreFilterPodTopologySpread\" from cycleState: not found" plugin="PodTopologySpread" pod="preemptionc64798a8-c9ef-4911-9a89-885ea274900e/preemptor-pod"
E1212 14:10:24.194204 1682349 schedule_one.go:146] "Status after running PostFilter plugins for pod" pod="preemptionc64798a8-c9ef-4911-9a89-885ea274900e/preemptor-pod" status=&{code:1 reasons:[preemption: running RemovePod on PreFilter plugin "PodTopologySpread": reading "PreFilterPodTopologySpread" from cycleState: not found] err:0xc005bfe4f0 failedPlugin:}
    preemption_test.go:1940: Pod preemptionc64798a8-c9ef-4911-9a89-885ea274900e/victim-pod is not getting evicted.
```

alculquicondor replied:
you still have to write into the `cycleState`. Did you try that?

chrishenzie replied:
Synced offline -- will leave this as a followup, see #114433

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -250,65 +270,86 @@ func TestAccessModeConflicts(t *testing.T) {
 
+// Because errors embedded in scheduler statuses cause issues during comparison.
+func deepEqualIgnoreError(got *framework.Status, want *framework.Status) bool {
+	return reflect.DeepEqual(got.Code(), want.Code()) && reflect.DeepEqual(got.Message(), want.Message())
```
Use `cmp.Diff` instead of `DeepEqual`

alculquicondor commented on pkg/scheduler/internal/cache/snapshot.go:
```diff
@@ -36,9 +36,9 @@ type Snapshot struct {
-	usedPVCSet sets.String
+	// usedPVCs contains a map of PVC names that have one or more scheduled pods using them,
+	// keyed in the format "namespace/name". Values are in the format "name".
+	usedPVCs   map[string]sets.Set[string]
```
```suggestion
	usedPVCs   map[string]int
```

is this enough?

chrishenzie commented on pkg/scheduler/internal/cache/cache.go:
```diff
@@ -310,8 +310,8 @@ func (cache *cacheImpl) updateNodeInfoSnapshotList(snapshot *Snapshot, updateAll
-				for key := range nodeInfo.PVCRefCounts {
-					snapshot.usedPVCSet.Insert(key)
+				for key, value := range nodeInfo.PVCRefs {
+					snapshot.usedPVCs[key] += value.Len()
```
It looks like I can revert this back to `PVCRefCount`, fixing that

alculquicondor commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -1607,3 +1609,355 @@ func TestPreferNominatedNode(t *testing.T) {
+			}
+
+			// Cleanup
+			pods = append(pods, preemptor)
+			testutils.CleanupPods(cs, t, pods)
+			if err := test.cleanup(); err != nil {
+				t.Errorf("Error cleaning up test: %v", err)
+			}
```
these should be called within `t.CleanUp`. Otherwise you might be leaving uncleaned items when tests fail.

alculquicondor commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -1607,3 +1609,355 @@ func TestPreferNominatedNode(t *testing.T) {
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := test.init(); err != nil {
+				t.Errorf("Error while initializing test: %v", err)
```
Fatalf

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -249,18 +271,33 @@ func TestAccessModeConflicts(t *testing.T) {
 	}
 
+	cmpOpts := []cmp.Option{
+		cmp.Comparer(func(s1 *framework.Status, s2 *framework.Status) bool {
```
`framework.Status` already implements `Equal`. You shouldn't need this.

chrishenzie replied:
I was encountering issues comparing statuses without this -- it gets hung-up on the pointer addresses:
```
--- FAIL: TestAccessModeConflicts (0.50s)
    --- FAIL: TestAccessModeConflicts/failed_to_get_PVC (0.10s)
        volume_restrictions_test.go:362: Unexpected Filter status (-want, +got):   (*framework.Status)(
            -   &⟪0xc0008f5a00⟫{
            -           code:    1,
            -           reasons: []string(⟪ptr:0xc000989890, len:1, cap:1⟫{`reading "PreFilterVolumeRestrictions" from cycleState: not found`}),
            -           err: &⟪0xc0008a6200⟫"fmt".wrapError{
            -                   msg: `reading "PreFilterVolumeRestrictions" from cycleState: not found`,
            -                   err: &⟪0xc0008181b0⟫"errors".errorString{s: "not found"},
            -           },
            -   },
            +   &⟪0xc0008f5e80⟫{
            +           code:    1,
            +           reasons: []string(⟪ptr:0xc000989d00, len:1, cap:1⟫{`reading "PreFilterVolumeRestrictions" from cycleState: not found`}),
            +           err: &⟪0xc0008a6420⟫"fmt".wrapError{
            +                   msg: `reading "PreFilterVolumeRestrictions" from cycleState: not found`,
            +                   err: &⟪0xc0008181b0⟫"errors".errorString{s: "not found"},
            +           },
            +   },
              )
```
I borrowed this pattern from here:
https://github.com/kubernetes/kubernetes/blob/v1.26.0/pkg/scheduler/schedule_one_test.go#L1484-L1488

kerthcet replied:
Can we compare the error log directly in `status.Equal()`?  Most of the time, we have a plain error.

chrishenzie replied:
Updated comparator to use `.Message()` and explicitly set test fields `preFilterWantStatus` and `wantStatus` to `nil` to clarify test expectations

kerthcet commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:
```diff
@@ -40,17 +41,60 @@ type VolumeRestrictions struct {
 )
 
+// ErrStatusKeyNotFound is returned when the plugin PreFilter state cannot be read.
+var ErrStatusKeyNotFound = fmt.Errorf("cannot read %q from cycleState", preFilterStateKey)
```
Nit: seems no need to export this error.

alculquicondor replied:
This should be enough

```suggestion
var errStatusKeyNotFound = fmt.Errorf("cannot read %q from cycleState", preFilterStateKey)
```

kerthcet reviewed (commented):
Looks like this is ready to merge now, anything left? @alculquicondor

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -252,47 +273,68 @@ func TestAccessModeConflicts(t *testing.T) {
 			enableReadWriteOncePod: true,
-			wantStatus:             framework.NewStatus(framework.UnschedulableAndUnresolvable, "persistentvolumeclaim \"claim-with-rwop\" not found"),
+			preFilterWantStatus:    framework.NewStatus(framework.UnschedulableAndUnresolvable, "persistentvolumeclaim \"claim-with-rwop-1\" not found"),
+			wantStatus:             framework.AsStatus(ErrStatusKeyNotFound),
```
Actually, what happens in the scheduler when preFilter returns UnschedulableAndUnresolvable?

IIUC, Filter wouldn't run. So having Filter run in the test is kind of useless, isn't it?

chrishenzie replied:
Updated to only run the `wantStatus` test if `preFilterWantStatus` is successful, which allowed me to get rid of `ErrStatusKeyNotFound`

alculquicondor commented on pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions_test.go:
```diff
@@ -301,9 +343,17 @@ func TestAccessModeConflicts(t *testing.T) {
+				t.Errorf("Unexpected PreFilter status (-want, +got): %s", diff)
+			}
+			// If PreFilter fails, then Filter will not run.
+			if test.preFilterWantStatus.IsSuccess() {
```
FYI @kidddddddddddddddddddddd regarding #114898

alculquicondor reviewed (commented):
/approve
/hold cancel
/lgtm

chrishenzie commented:
PRR approved, this is ready for merge. See https://github.com/kubernetes/kubernetes/pull/114494 for e2e test changes and feature gate bump to beta.

kerthcet commented:
/lgtm
ping @aojea for tests. We have other PR blocked by this one.

aojea commented:
> /lgtm ping @aojea for tests. We have other PR blocked by this one.

/approve
